interfaces, chain, refactor: Remove unused getTipLocator and incaccurate getActiveChainLocator

pablomartin4btc opened the pull request:
Remove `Chain::getTipLocator`, `Chain::GetLocator()`, and `Chain::getActiveChainLocator`:
- `Chain::getTipLocator` is no longer used.
- `Chain::GetLocator`, replaced its call by `GetLocator()`, which uses `LocatorEntries`, avoiding direct access to the chain itself (change suggested by l0rinc while reviewing this PR to maintain consistency with the overall refactoring).
- `Chain::getActiveChainLocator`, whose name was misleading, has functionality redundant with Chain::findBlock.
  - Additionally, the comment for getActiveChainLocator became inaccurate following changes in commit https://github.com/bitcoin/bitcoin/commit/ed470940cddbeb40425960d51cefeec4948febe4 (from PR #25717).


This is a [follow-up](https://github.com/bitcoin/bitcoin/pull/29652#issuecomment-3151665095) to #29652.

pablomartin4btc committed:
interfaces, chain, refactor: Remove unused getTipLocator

Also removed CChain::GetLocator() and replaced its call
with GetLocator() which uses LocatorEntries instead.

Co-authored-by: ryanofsky <ryan@ofsky.org>
Co-authored-by: l0rinc <l0rinc@users.noreply.github.com>

pablomartin4btc committed:
interfaces, chain, refactor: Remove inaccurate getActiveChainLocator

The getActiveChainLocator method name was misleading, and its functionality
duplicated `Chain::findBlock`. This commit removes the method and replaces
all its usages with direct `Chain::findBlock` calls.

Additionally, the comment of getActiveChainLocator has been outdated since
commit ed47094 from #25717.

Finally, in CWallet::ScanForWalletTransactions, the findBlock calls are now
unified into a single call at the start of the function.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>
Co-authored-by: Matias Furszyfer <mfurszy@protonmail.com>

stickies-v reviewed (approved):
ACK 9c0542f26055330b92967a0620fb9ae3f315eb19

Rationale makes sense, can't see any behaviour change.

stickies-v commented on src/wallet/wallet.cpp:
```diff
@@ -1858,7 +1858,8 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
+                    CBlockLocator loc;
+                    chain().findBlock(block_hash, FoundBlock().locator(loc));
 
                     if (!loc.IsNull()) {
```
note: my understanding is that (also prior to this PR), since `!block.IsNull()` and we're not doing any traversing to get the locator, `loc` should never be null. I verified this by adding an `Assume` and running the debug build test suite. Could potentially add that in here too if you think it makes sense?

<details>
<summary>git diff on 9c0542f260</summary>

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index d17156b2b0..bb56badd81 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1861,7 +1861,7 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
                     CBlockLocator loc;
                     chain().findBlock(block_hash, FoundBlock().locator(loc));
 
-                    if (!loc.IsNull()) {
+                    if (Assume(!loc.IsNull())) {
                         WalletLogPrintf("Saving scan progress %d.\n", block_height);
                         WalletBatch batch(GetDatabase());
                         batch.WriteBestBlock(loc);

```
</details>

furszy reviewed (commented):
I think you could implement the second commit cleaner, and slightly more performant, by unifying the `findBlock` calls (only performing one at the beginning):

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
--- a/src/wallet/wallet.cpp	(revision f6a52c6686b4d14aea2b05ca0413ad664bcf72e6)
+++ b/src/wallet/wallet.cpp	(date 1754934464756)
@@ -1837,9 +1837,13 @@
         chain().findBlock(block_hash, FoundBlock().inActiveChain(block_still_active).nextBlock(FoundBlock().inActiveChain(next_block).hash(next_block_hash)));
 
         if (fetch_block) {
-            // Read block data
+            // Read block data and locator if needed (the locator is usually null unless we need to save progress)
             CBlock block;
-            chain().findBlock(block_hash, FoundBlock().data(block));
+            CBlockLocator loc;
+            // Find block
+            FoundBlock found_block{FoundBlock().data(block)};
+            if (save_progress && next_interval) found_block.locator(loc);
+            chain().findBlock(block_hash, found_block);
 
             if (!block.IsNull()) {
                 LOCK(cs_wallet);
@@ -1857,14 +1861,10 @@
                 result.last_scanned_block = block_hash;
                 result.last_scanned_height = block_height;
 
-                if (save_progress && next_interval) {
-                    CBlockLocator loc = m_chain->getActiveChainLocator(block_hash);
-
-                    if (!loc.IsNull()) {
-                        WalletLogPrintf("Saving scan progress %d.\n", block_height);
-                        WalletBatch batch(GetDatabase());
-                        batch.WriteBestBlock(loc);
-                    }
+                if (!loc.IsNull()) {
+                    WalletLogPrintf("Saving scan progress %d.\n", block_height);
+                    WalletBatch batch(GetDatabase());
+                    batch.WriteBestBlock(loc);
                 }
             } else {
                 // could not scan block, keep scanning but record this block as the most recent failure
```

furszy commented:
> Thanks for the suggestion! I agree with it. I’ll prepare a separate follow-up commit to refactor that, so the current second commit stays minimal and clearly shows the intended change.

I don't really think adding an another commit just for this makes much sense. Feel free to take the code if you like it. We are talking about a simple refactoring here.

> When I first reviewed the code, I considered breaking down CWallet::ScanForWalletTransactions into smaller, more manageable pieces, which might be a worthwhile improvement in a separate PR. Please let me know if you also think that would be valuable.

it will depend on how you split it. Not sure if you can do much there; most of the "extra" code is to show and save progress. The rest seems very simple to me.

l0rinc commented on src/node/interfaces.cpp:
```diff
@@ -559,11 +559,6 @@ class ChainImpl : public Chain
-    CBlockLocator getTipLocator() override
-    {
-        LOCK(::cs_main);
-        return chainman().ActiveChain().GetLocator();
```
Now that we have one less `GetLocator` call, [the only remaining one](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2917) can also be inlined to `GetLocator(m_chain.Tip())` (we're already holding `cs_main`)

```patch
diff --git a/src/chain.cpp b/src/chain.cpp
index 82007a8a1e..4e2d1bf0ac 100644
--- a/src/chain.cpp
+++ b/src/chain.cpp
@@ -52,11 +52,6 @@ CBlockLocator GetLocator(const CBlockIndex* index)
     return CBlockLocator{LocatorEntries(index)};
 }
 
-CBlockLocator CChain::GetLocator() const
-{
-    return ::GetLocator(Tip());
-}
-
 const CBlockIndex *CChain::FindFork(const CBlockIndex *pindex) const {
     if (pindex == nullptr) {
         return nullptr;
diff --git a/src/chain.h b/src/chain.h
index c6d1640768..03dc801b98 100644
--- a/src/chain.h
+++ b/src/chain.h
@@ -467,9 +467,6 @@ public:
     /** Set/initialize a chain with a given tip. */
     void SetTip(CBlockIndex& block);
 
-    /** Return a CBlockLocator that refers to the tip in of this chain. */
-    CBlockLocator GetLocator() const;
-
     /** Find the last common block between this chain and a block index entry. */
     const CBlockIndex* FindFork(const CBlockIndex* pindex) const;
 
diff --git a/src/validation.cpp b/src/validation.cpp
index eb8bd2b57c..3dd8450dd4 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2941,7 +2941,7 @@ bool Chainstate::FlushStateToDisk(
     }
     if (full_flush_completed && m_chainman.m_options.signals) {
         // Update best block in wallet (so we can detect restored wallets).
-        m_chainman.m_options.signals->ChainStateFlushed(this->GetRole(), m_chain.GetLocator());
+        m_chainman.m_options.signals->ChainStateFlushed(this->GetRole(), GetLocator(m_chain.Tip()));
     }
     } catch (const std::runtime_error& e) {
         return FatalError(m_chainman.GetNotifications(), state, strprintf(_("System error while flushing: %s"), e.what()));
```

l0rinc replied:
👍, please resolve the comment

bc1cindy commented:
Concept ACK

Tested by building successfully and running wallet functional tests.

Good cleanup of unused interfaces. I lean towards furszy's suggestion of unifying the findBlock calls - it's cleaner and avoids redundant lookups for the same block. Simple enough to include here.

l0rinc reviewed (changes requested):
Only had time to review 9a4849f17b1b56a40795c2bc5e953d44959d5e80

pablomartin4btc commented:
<ins>_Updates_</ins>:
- Addressed feedback from @furszy, incorporating [suggested refactoring](https://github.com/bitcoin/bitcoin/pull/33169#pullrequestreview-3107022031).
- Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/33169#discussion_r2268362028) from @l0rinc, removing `Chain::GetLocator()` function and replacing its call to be consistent with the overall refactoring done in the PR.

l0rinc commented on src/wallet/wallet.cpp:
```diff
@@ -1837,9 +1837,13 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
+            // Read block data and locator if needed (the locator is usually null unless we need to save progress)
             CBlock block;
-            chain().findBlock(block_hash, FoundBlock().data(block));
+            CBlockLocator loc;
```
could we move this closer to its first usage instead of announcing it early?

pablomartin4btc replied:
Sorry, which bit specifically? I can do if I have to retouch.

l0rinc replied:
the locator isn't needed for the next line, we could narrow its scope by moving it closer to its first usage

l0rinc commented on src/wallet/wallet.cpp:
```diff
@@ -1837,9 +1837,13 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
+            CBlockLocator loc;
+            // Find block
+            FoundBlock found_block{FoundBlock().data(block)};
+            if (save_progress && next_interval) found_block.locator(loc);
```
not sure I understand why we've moved this out of the condition.
Could we split this into two commits, one removing the need for `getActiveChainLocator`, something like:
```patch
diff --git a/src/interfaces/chain.h b/src/interfaces/chain.h
index f45cb8e598..82e626e35d 100644
--- a/src/interfaces/chain.h
+++ b/src/interfaces/chain.h
@@ -143,10 +143,6 @@ public:
     //! pruned), and contains transactions.
     virtual bool haveBlockOnDisk(int height) = 0;
 
-    //! Return a locator that refers to a block in the active chain.
-    //! If specified block is not in the active chain, return locator for the latest ancestor that is in the chain.
-    virtual CBlockLocator getActiveChainLocator(const uint256& block_hash) = 0;
-
     //! Return height of the highest block on chain in common with the locator,
     //! which will either be the original block used to create the locator,
     //! or one of its ancestors.
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index d7ada5ae75..106c961173 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -559,12 +559,6 @@ public:
         const CBlockIndex* block{chainman().ActiveChain()[height]};
         return block && ((block->nStatus & BLOCK_HAVE_DATA) != 0) && block->nTx > 0;
     }
-    CBlockLocator getActiveChainLocator(const uint256& block_hash) override
-    {
-        LOCK(::cs_main);
-        const CBlockIndex* index = chainman().m_blockman.LookupBlockIndex(block_hash);
-        return GetLocator(index);
-    }
     std::optional<int> findLocatorFork(const CBlockLocator& locator) override
     {
         LOCK(::cs_main);
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 21de98862b..fe82383ebd 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1839,7 +1839,8 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
         if (fetch_block) {
             // Read block data
             CBlock block;
-            chain().findBlock(block_hash, FoundBlock().data(block));
+            auto found_block{FoundBlock().data(block)};
+            chain().findBlock(block_hash, found_block);
 
             if (!block.IsNull()) {
                 LOCK(cs_wallet);
@@ -1858,7 +1859,8 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
                 result.last_scanned_height = block_height;
 
                 if (save_progress && next_interval) {
-                    CBlockLocator loc = m_chain->getActiveChainLocator(block_hash);
+                    CBlockLocator loc;
+                    found_block.locator(loc);
 
                     if (!loc.IsNull()) {
                         WalletLogPrintf("Saving scan progress %d.\n", block_height);
```

Which is easy to understand and a second one, moving the locator assignment before the block null check, i.e. something like:
```patch
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index fe82383ebd..8f2bbc23c9 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1837,9 +1837,12 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
         chain().findBlock(block_hash, FoundBlock().inActiveChain(block_still_active).nextBlock(FoundBlock().inActiveChain(next_block).hash(next_block_hash)));
 
         if (fetch_block) {
-            // Read block data
             CBlock block;
             auto found_block{FoundBlock().data(block)};
+
+            CBlockLocator loc;
+            if (save_progress && next_interval) found_block.locator(loc);
+
             chain().findBlock(block_hash, found_block);
 
             if (!block.IsNull()) {
@@ -1858,15 +1861,10 @@ CWallet::ScanResult CWallet::ScanForWalletTransactions(const uint256& start_bloc
                 result.last_scanned_block = block_hash;
                 result.last_scanned_height = block_height;
 
-                if (save_progress && next_interval) {
-                    CBlockLocator loc;
-                    found_block.locator(loc);
-
-                    if (!loc.IsNull()) {
-                        WalletLogPrintf("Saving scan progress %d.\n", block_height);
-                        WalletBatch batch(GetDatabase());
-                        batch.WriteBestBlock(loc);
-                    }
+                if (!loc.IsNull()) {
+                    WalletLogPrintf("Saving scan progress %d.\n", block_height);
+                    WalletBatch batch(GetDatabase());
+                    batch.WriteBestBlock(loc);
                 }
             } else {
                 // could not scan block, keep scanning but record this block as the most recent failure
```

Which I'm not yet sure I fully understand. The commit message doesn't help me, but after the split the commit messages could explain the separately with more context, maybe that would help.

pablomartin4btc replied:
Ok, I'll try that. Thanks!

l0rinc reviewed (changes requested):
Concept ACK

The last commit is could be split into two, the commit message could explain higher level of what's happening exactly.

furszy reviewed (commented):
ACK 2b00030af84efd50cfc60125db4ae49da6b7aa9a

stickies-v reviewed (approved):
ACK 2b00030af84efd50cfc60125db4ae49da6b7aa9a

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33169/commits/2b00030af84efd50cfc60125db4ae49da6b7aa9a

achow101 commented:
ACK 2b00030af84efd50cfc60125db4ae49da6b7aa9a
